table/block_based/block_based_table_iterator.cc

```diff
@@ -557,5 +700,119 @@ void BlockBasedTableIterator::FindReadAheadSizeUpperBound() {
 
   block_prefetcher_.SetUpperBoundOffset(start_offset +
                                         total_bytes_till_upper_bound);
+
+  if (getenv("Print")) {
+    printf("FindReadAheadSizeUpperBound:End\n");
+  }
+}
+
+void BlockBasedTableIterator::BlockCacheLookupForReadAheadSize(
+    size_t readahead_size, size_t& updated_readahead_size) {
+  if (getenv("Print")) {
+    printf("BlockCacheLookupForReadAheadSize\n");
+  }
+  ClearBlockHandles();
+  /*
+    Akanksha Note:
+    1. Call block based table reader to lookup the data block in the cache (from
+    offset till readahead_size).
+    2. Iterate over index_block.
+    3. If block is in cache, pin the block.
+    4. If block is not in the cache, add place holder in cache.
+    5. Add the handle to the vector.
+    6. Iterate over the vector backwards to find the last cache miss and update
+    readahead_size.
+    4. Update readahead_size till last hit within readahead_size.
+  */
+  updated_readahead_size = readahead_size;
+
+  // readahead_cache_lookup_ can be false after Seek, if after Seek or Next
+  // there is Prev or any other operation.
+  if (!readahead_cache_lookup_) {
+    return;
+  }
+
+  size_t current_readahead_size = 0;
+  size_t footer = table_->get_rep()->footer.GetBlockTrailerSize();
+
+  BlockHandleInfo* block_handle_info = new BlockHandleInfo();
```

Comment:
Can we avoid these individual allocations?

Reply:
You mean the first allocation outside the while loop?
That would add extra checks in the while loop as for first block - we don't do the lookup and don't include it to readahead_size. So I have to add the if statement for the first block in while loop.

Reply:
No, I mean can we define ```block_handles_``` as ```std::deque<BlockHandleInfo>``` and use ```emplace_back```? A ```deque``` element is anyway dynamically allocated, so we can avoid the extra ```BlockHandleInfo``` allocation.